### PR TITLE
chore(chart): Updated release logic to use SemVer

### LINKED ISSUE #4248: Support the Semver2 standard
### Description

**What problem are you trying to solve?**

Attempts to get the latest tags utilizing semver ranges via Helm or FluxCD's HelmRelease resource will fail.

This is caused by all versions of karpenter helm chart being prefixed with `v`.
In fact, if you look at public.ecr.aws/karpenter, you'll see the following tags.

```shell
➜  tmp TOKEN=$(curl -k https://public.ecr.aws/token/ | jq -r '.token')

➜  tmp curl -k -H "Authorization: Bearer $TOKEN" https://public.ecr.aws/v2/karpenter/karpenter/tags/list  | jq . | head
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 42196    0 42196    0     0  16310      0 --:--:--  0:00:02 --:--:-- 16304
{
  "name": "karpenter/karpenter",
  "tags": [
    "v0-69bd00df79db89a1171ef950537230e5493cd60e",
    "v0-31de5dcf837359033cf40b72db989e3b96fa46f9",
    "v0-ba600d8038747224dd55d89156e4bd65512a6000",
    "20220911",
    "20220912",
    "20220910",
    "v0-dcbb2d8de30e7c4f13670cf3bffd3d47a324d662",
```

Due to this version format, the helm pull command will fail to get the latest tags using semver range.

```shell
➜  tmp helm pull oci://public.ecr.aws/karpenter/karpenter --version="0.*" --debug 
Error: Unable to locate any tags in provided repository: oci://public.ecr.aws/karpenter/karpenter
helm.go:84: [debug] Unable to locate any tags in provided repository: oci://public.ecr.aws/karpenter/karpenter
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).getOciURI
	helm.sh/helm/v3/pkg/downloader/chart_downloader.go:159
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).ResolveChartVersion
	helm.sh/helm/v3/pkg/downloader/chart_downloader.go:199
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).DownloadTo
	helm.sh/helm/v3/pkg/downloader/chart_downloader.go:90
helm.sh/helm/v3/pkg/action.(*Pull).Run
	helm.sh/helm/v3/pkg/action/pull.go:131
main.newPullCmd.func2
	helm.sh/helm/v3/cmd/helm/pull.go:74
github.com/spf13/cobra.(*Command).execute
	github.com/spf13/cobra@v1.6.1/command.go:916
github.com/spf13/cobra.(*Command).ExecuteC
	github.com/spf13/cobra@v1.6.1/command.go:1044
github.com/spf13/cobra.(*Command).Execute
	github.com/spf13/cobra@v1.6.1/command.go:968
main.main
	helm.sh/helm/v3/cmd/helm/helm.go:83
runtime.main
	runtime/proc.go:250
runtime.goexit
	runtime/asm_amd64.s:1598
```

The tag list we get via the `tags/list` endpoint is 1576,

```shell
registryTags - []string len: 1576, cap: 2048, ["v0-69bd00df79db89a1171ef950537230e5493cd60e","v0-31de5dcf837359033cf40b72db989e3b96fa46f9","v0-ba600d8038747224dd55d89156e4bd65512a6000","20220911","20220912","20220910","v0-dcbb2d8de30e7c4f13670cf3bffd3d47a324d662","v0-49fac0ca5a5d4b98e7e5801df2890c9e044fe99c","v0-80e60d8a9a737b2a8dc0aaf43140f9928db8c337","v0-49a318ecfbc59e79d69115ac8dc1c9429616b671","v0-5fc55df4f570d8ccbb6b503dca37bc35d6d2d405","v0-a255b82814e6c94e0e933d23ed5cd76126408770","v0-cdf4fa69b339742ae0429267845fa16918627338","v0-9cbaa76d3c6c0ecbc23d46ebc1693bb1d2ad5418","v0-6217d62d1d199bd61e252b9a55f9539430efaeff","20221116","v0-3ead1d92f7b827a30b851929e3832761cde9641e","v0-b828a83db3fce5c5ad7c28581d68cd6c7aa6314d","v0-36ebd55249f736f2bb5c13777dc0a5656e338639","v0-111c114b6dfe7e8403a2c16e98d656ba6ad807c3","v0-62b1069f25b30280d22e8aa99c4d7695b4fb38e7","v0-558393be817833ed7b4b9359259e76027bf6e522","v0-33bfe9e960dd65c808ec410256b9b170a43e9e53","20220808","20220807","20220806","v0-b1c1a9da8b34a68801e5c6b5c15fc24054c09983","20220923","v0-1e6c5d9b3bf87c21eaaeb7a78248c2db407dee96","v0-8e92f731f0bd16894e949fd0254f7c9631579360","20221107","v0-30ce02a5a20398bf9b624fdd899a48f0987c27b9","v0-d4ec73075788514eaa7a6d6ea8957a7204ca5cec","v0-5a49e01bba57fa1ba82733b87853c9644ed5e56b","v0-582ca837b7b976689ca33b100bb531aa9de6097f","v0-c5a2044907258346afc879f7bf7867f9e15a1bb4","v0-6bdd21387902c14cb4f7be252558fbffbe654a6b","v0-fa3a89accab4f7f7d3973078f1f513ec85022674","v0-471148a26513e25b02aa3f32a51116490ed6feab","v0-4ff98b06de9d27a0ec8cd7ff6a5d65b5050d90bf","v0-109295bb3be995f1487afc402298bf278f768433","v0-5333fb247531c60038d5b3b9aae45a406b257b28","v0-5661ebef509dbfca0bc2e8529311e2ebcdc9adb5","v0-2a93373ed6d3f8717b336060f78a699655a8be3a","v0-a35dd1e56418e9ac92aaff15096197ab748450fb","v0-470aa83476d121f9215d7a209d6725b2d0e11738","v0-125b96fce06ab4e10b8f56f7c6cca5cfc9eb4556","v0-a34da16c8e3b60ba6c380b7a03248d3557ea24b9","v0-2e34488038bdf346dd407051c617f2954a517f66","v0-0b25c60640757741d733c45eb9bbc7160a7d62b1","v0-4300212384bb7a051d82e11b61d1b0831e33868c","v0-117f007c50646d57825387d47ebcea9b10f769b0","v0-698f22f5c07ba1cef4ffb5d80dee482d37b717c0","v0-7657e823b38af57393ddf1ede4447055779787ca","v0-e9b6a72e003b81da3ee51ce7850947b3ea19391a","v0-d43c9b2cac2b4ae98199e80e4660244646690284","v0-2f949e5a1be3877fb6a932598054638f406164e2","v0-448cb6662c5d53e5cfd6751c8da696ac9faca9eb","v0-958b40484560affe9883c838d80cc8891f212f91","v0-f2a49d45b76ac113e74adbf6a7cb778083a71b9a","v0.18.0","v0-1eeb010f80397e2f8ff412a377f9b18a782bc681","v0-717adebe9c6aa6294a3f52ae34f097af90a2cddf","v0-653b456f6b0156009f82db3389d46dd24802bf3c",...+1512 more]
```

However, since HELM's implementation looks like this, none of the valid semvers are filtered out.

```go
var tagVersions []*semver.Version
for _, tag := range registryTags {
	// Change underscore (_) back to plus (+) for Helm
	// See https://github.com/helm/helm/issues/10166
	tagVersion, err := semver.StrictNewVersion(strings.ReplaceAll(tag, "_", "+"))
	if err == nil {
		tagVersions = append(tagVersions, tagVersion)
	}
}
```

"Invalid characters in version."

```shell
tagVersions - []*github.com/Masterminds/semver/v3.Version len: 0, cap: 0, nil
```

**How important is this feature to you?**

https://helm.sh/docs/topics/charts/#charts-and-versioning
> Every chart must have a version number. A version must follow the [SemVer 2](https://semver.org/spec/v2.0.0.html) standard. Unlike Helm Classic, Helm v2 and later uses version numbers as release markers. Packages in repositories are identified by name plus version.

To keep track of the latest tags using semver ranges, why not use the available semvers for the StrictNewVersion method and upload them to the public ecr repo?

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by stevehipwell**
Fixes #4248

**Description**
This PR changes the chart release implementation to use a valid SemVer version (`x.y.z`) for the chart version by removing the tag's `v` prefix; this matches the modification to _Chart.yaml_ made in the stable branch PR. This PR also updates the documentation to support this change.

Once this has been merged and the first release has been cut with this code it'd make sense to also create SemVer tags for the other versions supporting `v1beta1` (`0.32.0`, `0.32.1`, `0.32.2`, `0.32.3`, `0.32.4`, `0.32.5`, `0.32.6`, `0.33.0`, `0.33.1`, `0.33.2` & `0.34.0`). The chart packages would need to be pulled and the chart version modified to remove the `v` prefix before being pushed back up.

**How was this change tested?**
Ad-hoc

**Does this change impact docs?**
- [x] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by stevehipwell** [comment]
CC @jonathan-innis

**Comment by jonathan-innis** [comment]
This looks like it adds the semver change combined with the artifacthub change. Can we break those changes apart?

**Comment by stevehipwell** [comment]
@jonathan-innis it could be broken apart but I'm not sure of the value in doing so? Both changes are release process changes which make sense to be tested together. The actual Artifact Hub change is ignored until the org has been configured, the repos adopted or recreated and the config file content added.

That said, I'll split it up if you'd like.

**Comment by jonathan-innis** [comment]
> That said, I'll split it up if you'd like

Yeah, I get that it's really not that difficult to add one way or the other. Mainly in the spirit of making sure all of the PRs accurately describe what they are doing and the commit squash history correctly captures everything

File: hack/release/common.sh [hunk 1]
```diff
@@ -127,13 +127,17 @@ publishHelmChart() {
...
     cd charts
+    [[ -s "${AH_CONFIG_FILE_NAME}" ]] && oras push "${RELEASE_REPO}:artifacthub.io" --config /dev/null:application/vnd.cncf.artifacthub.config.v1+yaml "${AH_CONFIG_FILE_NAME}:application/vnd.cncf.artifacthub.repository-metadata.layer.v1.yaml"
     helm dependency update "${CHART_NAME}"
     helm lint "${CHART_NAME}"
-    helm package "${CHART_NAME}" --version "$HELM_CHART_VERSION"
+    helm package "${CHART_NAME}" --version "${HELM_CHART_VERSION}"
     helm push "${HELM_CHART_FILE_NAME}" "oci://${RELEASE_REPO}"
     rm "${HELM_CHART_FILE_NAME}"
+    digest="$(crane digest "${RELEASE_REPO}:${HELM_CHART_VERSION}")"
+    cosign sign --yes "${RELEASE_REPO}:${HELM_CHART_VERSION}@${digest}"
```
Comment (jonathan-innis) [inline hack/release/common.sh:140]: Nice! I've been wanting to do this for a while! Can we break this out into a separate PR as well? Super excited that we will be cosigning these now!

As an aside, do you think there is value in adding a note on valdiating the cosign signature as part of the helm chart install in the docs? Or at least to mention it somewhere.

File: hack/toolchain.sh [hunk 2]
```diff
@@ -23,6 +23,8 @@ tools() {
     go install github.com/onsi/ginkgo/v2/ginkgo@latest
     go install github.com/rhysd/actionlint/cmd/actionlint@latest
     go install github.com/mattn/goveralls@latest
+    go install github.com/google/go-containerregistry/cmd/crane@latest
```
Comment (jonathan-innis) [inline hack/toolchain.sh:26]: Seems like we won't need these without the artifacthub pushing or the cosigning

File: website/content/en/preview/getting-started/migrating-from-cas/_index.md [hunk 3]
```diff
@@ -92,10 +92,10 @@ One for your Karpenter node role and one for your existing node group.
 First set the Karpenter release you want to deploy.
 
 ```bash
-export KARPENTER_VERSION={{< param "latest_release_version" >}}
+export KARPENTER_VERSION="{{< param "latest_release_version" >}}"
```
Comment (jonathan-innis) [inline website/content/en/preview/getting-started/migrating-from-cas/_index.md:95]: Does it make sense to just change this version to not have the `v` and then append the `v` onto the version when we need it, like when we are pulling raw files from Github?

**Review by jonathan-innis — COMMENTED** [review]
Few comments. Nice work!

File: website/content/en/preview/getting-started/migrating-from-cas/_index.md (hunk 3 above)
Comment (stevehipwell) [reply website/content/en/preview/getting-started/migrating-from-cas/_index.md:95]: That was my original plan, so I'm happy to pivot to this.

File: hack/toolchain.sh (hunk 2 above)
Comment (stevehipwell) [reply hack/toolchain.sh:26]: Crane is needed for cosigning, oras is needed for the artifact hub metadata.

File: hack/release/common.sh (hunk 1 above)
Comment (stevehipwell) [reply hack/release/common.sh:140]: Yes, I think this would be a useful addition, maybe an optional sub-step?

**Comment by stevehipwell** [comment]
> Yeah, I get that it's really not that difficult to add one way or the other. Mainly in the spirit of making sure all of the PRs accurately describe what they are doing and the commit squash history correctly captures everything

@jonathan-innis I'm of a similar opinion; but the reality of a non-contributor contribution is high friction, so bunching similar changes into a single PR makes much better use of available time.

Shall we get all of the changes into this PR and agreed on to cut the initial load? Once the changes have been agreed on in principal I can split it up for testing if required.

**Comment by jonathan-innis** [comment]
> Shall we get all of the changes into this PR and agreed on to cut the initial load

The changes look good to me as-is, aside from changing the format of the Karpenter version parameter. I'd still opt for separate PRs, but I don't think there will be a ton of back-and-forth since these changes already look alright

**Comment by stevehipwell** [comment]
@jonathan-innis I'm planning on changing the meaning of `KARPENTER_VERSION` in the docs and all of the scripts for consistency, so there will be a bit of reviewing required. I'll make the change here and then split the PR up once you're happy in principal. I'm thinking 3 PRs; 1 for the chart SemVer & docs, 1 for signing the chart artifact & 1 for managing the Artifact Hub config.

**Comment by stevehipwell** [comment]
@jonathan-innis this should be all of the changes. I spent a long time looking at the scripts in `hack` but on reflection I decided that I was going to need to effectively re-write them for very little real world benefit.

**Comment by jonathan-innis** [comment]
> this should be all of the changes

Sounds good. Do you mind breaking up the changes for the aritfactrepo and the cosigning like we discussed originally so we can stage that in a different PR and it makes it easier for me to see which changes are necessary here?

File: website/content/en/preview/getting-started/migrating-from-cas/scripts/step05-tag-subnets.sh [hunk 4]
```diff
@@ -1,6 +1,6 @@
-for NODEGROUP in $(aws eks list-nodegroups --cluster-name ${CLUSTER_NAME} \
-    --query 'nodegroups' --output text); do aws ec2 create-tags \
+for NODEGROUP in $(aws eks list-nodegroups --cluster-name "${CLUSTER_NAME}" --query 'nodegroups' --output text); do
+    aws ec2 create-tags \
         --tags "Key=karpenter.sh/discovery,Value=${CLUSTER_NAME}" \
-        --resources $(aws eks describe-nodegroup --cluster-name ${CLUSTER_NAME} \
-        --nodegroup-name $NODEGROUP --query 'nodegroup.subnets' --output text )
+        --resources "$(aws eks describe-nodegroup --cluster-name "${CLUSTER_NAME}" \
```
Comment (jonathan-innis) [inline website/content/en/preview/getting-started/migrating-from-cas/scripts/step05-tag-subnets.sh:4]: Does quoting work inside of quoting here?

File: website/content/en/preview/getting-started/migrating-from-cas/scripts/step01-env.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 AWS_PARTITION="aws" # if you are not using standard partitions, you may need to configure to aws-cn / aws-us-gov
 AWS_REGION="$(aws configure list | grep region | tr -s " " | cut -d" " -f3)"
-OIDC_ENDPOINT="$(aws eks describe-cluster --name ${CLUSTER_NAME} \
+OIDC_ENDPOINT="$(aws eks describe-cluster --name "${CLUSTER_NAME}" \
```
Comment (jonathan-innis) [inline website/content/en/preview/getting-started/migrating-from-cas/scripts/step01-env.sh:3]: Same question here

File: website/content/en/preview/getting-started/getting-started-with-karpenter/scripts/step02-create-cluster.sh [hunk 6]
```diff
@@ -67,7 +67,7 @@ addons:
 #  - namespace: "${KARPENTER_NAMESPACE}"
 EOF
 
-export CLUSTER_ENDPOINT="$(aws eks describe-cluster --name ${CLUSTER_NAME} --query "cluster.endpoint" --output text)"
+export CLUSTER_ENDPOINT="$(aws eks describe-cluster --name "${CLUSTER_NAME}" --query "cluster.endpoint" --output text)"
```
Comment (jonathan-innis) [inline website/content/en/preview/getting-started/getting-started-with-karpenter/scripts/step02-create-cluster.sh:70]: Same question here

File: hack/toolchain.sh (hunk 2 above)
Comment (jonathan-innis) [reply hack/toolchain.sh:26]: I think if we break out those changes into separate PRs like we discussed, then we don't need the dependencies here in the context of this PR

File: hack/release/common.sh (hunk 1 above)
Comment (jonathan-innis) [reply hack/release/common.sh:140]: > maybe an optional sub-step

Yeah, definitely an optional sub-step, not required at all! I was mostly just thinking if you were going to be opening a PR to be handling the cosigning that it might be nice to add the verification piece there with that change

File: website/content/en/preview/getting-started/migrating-from-cas/scripts/step05-tag-subnets.sh (hunk 4 above)
Comment (stevehipwell) [reply website/content/en/preview/getting-started/migrating-from-cas/scripts/step05-tag-subnets.sh:4]: Yes, it's best practice to always quote strings in bash.

**Comment by stevehipwell** [comment]
@jonathan-innis this PR should now be scoped to only the SemVer change.

**Review by jonathan-innis — DISMISSED** [review]
LGTM 🚀

Looks like it just needs a rebase

**Comment by stevehipwell** [comment]
@jonathan-innis this has been rebased and is ready to be merged.

**Comment by jonathan-innis** [comment]
A good question that @njtran raised here: Do we want to remove the prefix from the image tags as well? I've seen both in K8s but in the spirit of getting our tagging in general in a proper semantic versioning scheme, does it make sense that both of these get changed together?

**Comment by jonathan-innis** [comment]
One question that got raised here: If we are changing the version of our Helm chart to be semver-compliant, we might as well make the change to our images as well. Thoughts on having our tags in our Git repo being v-prefixed but having the assets (images and helm chart remove the prefix) @stevehipwell?

**Comment by stevehipwell** [comment]
> One question that got raised here: If we are changing the version of our Helm chart to be semver-compliant, we might as well make the change to our images as well. Thoughts on having our tags in our Git repo being v-prefixed but having the assets (images and helm chart remove the prefix) @stevehipwell?

@jonathan-innis this PR is fixing the fact that Helm explicitly linked the OCI tag to the chart SemVer; OCI images don't have this constraint and both patterns are common. Given this the OCI image tag format is related but isn't logically part of the same changes and the fact that this PR has already being split up to focus on a single change I think it should be discussed in a new issue and if implemented done so in a new PR.

FWIW I personally prefer using the SemVer version for an OCI image tag instead of replicating the git tag `v` prefix.

**Comment by jonathan-innis** [comment]
> SemVer version for an OCI image tag instead of replicating the git tag v prefix

Agreed. I think we should make a change in both. Happy to approve this change. Are you open to making the PR to update the OCI tag that we use for images as well?

**Comment by stevehipwell** [comment]
> Are you open to making the PR to update the OCI tag that we use for images as well?

@jonathan-innis yes I'm happy to do that too.

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Comment by billylaing** [comment]
Hello! Thanks so much for updating the tagging for Karpenter! When can we expect to see the new tags pushed out to `oci://public.ecr.aws/karpenter/karpenter`?

**Comment by stevehipwell** [comment]
@billylaing check out the [0.35.0](https://github.com/kubernetes-sigs/karpenter/releases/tag/v0.35.0) release.

**Comment by billylaing** [comment]
Thanks! I just got Karpenter running in our cluster yesterday. Really cool so far.
